### PR TITLE
MAINT: Remove all nose testing support.

### PR DESCRIPTION
**Pull request by charris**
NumPy switched to using pytest in 2018 and nose has been unmaintained for many years. We have kept NumPy's nose support to avoid breaking downstream projects who were might have been using it and not yet switched to pytest or some other testing framework. With the arrival of Python 3.12, unpatched nose will raise an error. It is time to move on.

Decorators removed

- raises
- slow
- setastest
- skipif
- knownfailif
- deprecated
- parametrize
- _needs_refcount

These are not to be confused with pytest versions with similar names, e.g., pytest.mark.slow, pytest.mark.skipif, pytest.mark.parametrize.

Functions removed

- Tester
- import_nose
- run_module_suite

### CODE REVIEW

File: numpy/typing/tests/data/reveal/testing.pyi [hunk 1]
```diff
@@ -113,7 +113,6 @@ reveal_type(np.testing.rundocs())  # E: None
 reveal_type(np.testing.rundocs("test.py"))  # E: None
 reveal_type(np.testing.rundocs(Path("test.py"), raise_on_error=True))  # E: None
 
-@np.testing.raises(RuntimeError, RuntimeWarning)
```
Comment (charris) [inline numpy/typing/tests/data/reveal/testing.pyi:116]: @BvB93 Removing the decorator doesn't seem to bother the tests.

**Comment by charris** [comment]
The pyinstaller error looks unrelated. I wonder if the new version 5.7.0 released Dec 4 is the cause.

**Comment by mattip** [comment]
The error is
```
Traceback (most recent call last):
  File "numpy/_pyinstaller/pyinstaller-smoke.py", line 14, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "numpy/__init__.py", line 143, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "numpy/core/__init__.py", line 100, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "numpy/core/_add_newdocs.py", line 4970, in <module>
  File "numpy/core/function_base.py", line 529, in add_newdoc
ModuleNotFoundError: No module named 'numpy.core._multiarray_tests'
```

and seems to be related to this line https://github.com/numpy/numpy/blob/e2fccd939e12ba33e90f663fe158d3e7566d031e/numpy/core/_add_newdocs.py#L4970

**Comment by seberg** [comment]
It seems `raises` and `run_module_suite` are main names available through `np.testing.` that are removed.  We could add a brief:
```
def __getattr__(self, name):
    # These were removed in 1.25
    if name == "raises" or name == "run_module_suite":
        raise AttributeError(
            "numpy.testing.raises and numpy.random.run_module_suite wrapped "
            "`nose` and has been removed. We recommend switching to `pytest`.")
```
But, I am happy either way.  The typing test looks like the test for the decorator typing itself.

**Comment by mattip** [comment]
Maybe the error is due to a circular import somehow when using the pyinstaller hook?

**Comment by seberg** [comment]
There is a `nose` "excludedimports" in pyinstaller we can delete, but that seems like a tiny cleanup that shouldn't cause a failure.  I restarted the test out of curiosity, but no good idea, the hook seems to collect all dynamic libs in `numpy`, so missing the module seems strange.

**Comment by charris** [comment]
Hmm... the noseclasses file imported from `_multiarray_tests`, maybe we need another import somewhere.

**Comment by charris** [comment]
Importing `_multiarray_tests` in `_add_newdocs.py` fixes the missing module problem. The question is if that is where we want to import it.

There are a number of profiling errors of the form:
```
2023-01-19T21:08:46.8102125Z libgcov profiling error:/home/runner/work/numpy/numpy/build/temp.linux-x86_64-3.9/numpy/core/src/common/npy_hashtable.gcda:overwriting an existing profile data with a different timestamp
```
Which I don't think should be related unless removing nose has exposed a potential profiling problem.

**Comment by seberg** [comment]
Ohhh, the tests are the issue...  I suspect you can also add it to:
```
hiddenimports = ['numpy.core._dtype_ctypes']
```
in the pyinstaller hook file.  Not sure about the other thing, but I doubt it is related.

**Comment by charris** [comment]
> Ohhh, the tests are the issue... I suspect you can also add it to:

I considered that, made me wonder if pyinstaller was working correctly before. In the absence of complaints I decided to leave it as is, especially as I don't really understand it :)

**Comment by charris** [comment]
We should figure out how to deprecate `ModuleDeprecationWarning`. I wonder if deprecation warnings work properly in exceptions.

**Comment by charris** [comment]
Squashed and release note added.

**Comment by charris** [comment]
My guess is that things would work fine if the `_pyinstaller` directory were moved down below `conftest.py`, or even if the import were added to `_pytesttester.py`.

**Review by seberg — APPROVED** [review]
Thanks LGTM!  Could add formatting to the release note, but doesn't matter.

I am very sure the `_multiarray_tests` should be in `hiddenimport`, but I am happy to follow up on that myself.  Nothing in NumPy needs/uses it (so there is no way pyinstaller can find it by searching for `import` statements).
But the tests need it and we want to be able to run them, so we should make sure it gets packaged.  Having a stray import in the code works, but seems less clean.

**Comment by mattip** [comment]
Thanks @charris

**Pull request merged**